src/librustc_mir/build/expr/stmt.rs

```diff
@@ -79,14 +79,31 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 block.unit()
             }
             ExprKind::Continue { label } => {
-                this.break_or_continue(expr_span, label, block,
-                                       |loop_scope| loop_scope.continue_block)
+                let LoopScope { continue_block, extent, .. } =
+                    *this.find_loop_scope(expr_span, label);
+                this.break_or_continue(expr_span, block, continue_block, extent)
             }
-            ExprKind::Break { label } => {
-                this.break_or_continue(expr_span, label, block, |loop_scope| {
-                    loop_scope.might_break = true;
-                    loop_scope.break_block
-                })
+            ExprKind::Break { label, mut value } => {
+                let (break_block, extent, destination) = {
+                    let LoopScope {
+                        ref mut might_break,
+                        break_block,
+                        extent,
+                        ref destination,
+                        ..
+                    } = *this.find_loop_scope(expr_span, label);
+                    *might_break = true;
+                    // FIXME unnecessary clone
+                    (break_block, extent, destination.clone())
+                };
+                if let Some(dest) = destination {
+                    if let Some(value) = value.take() {
```

Comment:
The reason why this code is so awkward is that we can't keep `this` borrowed when we get to further calls like `this.into`. (`find_loop_scope` borrows `&mut self` and that's tied to the returned `&mut LoopScope`).
Also, the first branch already takes care of the returned block (it uses `unpack!`; `into` returns `BlockAnd<()>`) while the second branch doesn't return a block (`push_assign_unit` returns `()`).

`.take()` is a remnant though, my bad.